File: solutions/c/sieve/1/sieve.c
#include "sieve.h"

uint32_t sieve(uint32_t limit, uint32_t *primes, size_t max_primes)
{
    uint32_t numbers[limit + 1];
    uint32_t i;
    uint32_t j;
    numbers[0] = 0;
    numbers[1] = 0;
    for (i = 2; i <= limit; i++)
        numbers[i] = i;
    for (i = 2; i <= limit; i++)
    {
        for (j = 2; i * j <= limit; j++)
            numbers[i * j] = 0;
    }
    j = 0;
    for (i = 0; i <= limit; i++)
    {
        if (numbers[i] != 0) {
            primes[j] = numbers[i];
            j++;
            if (j == max_primes)
                return (j);
        }
    }
    return (j);
}

File: solutions/fortran/sieve/1/sieve.f90
module sieve
  implicit none

contains

  function primes(limit) result(array)
    integer, intent(in) :: limit
    integer, allocatable :: array(:)
    logical, allocatable :: sieve(:)
    integer :: i, j, count

    ! Initialize sieve array with .true. (indicating prime) for all numbers
    allocate(sieve(limit))
    sieve = .true.
    
    ! Set sieve(1) to .false. because 1 is not a prime number
    sieve(1) = .false.

    ! Perform sieve of Eratosthenes
    do i = 2, int(sqrt(real(limit)))
      if (sieve(i)) then
        do j = i*i, limit, i
          sieve(j) = .false.
        end do
      end if
    end do

    ! Count the number of primes
    count = 0
    do i = 2, limit
      if (sieve(i)) count = count + 1
    end do

    ! Allocate array to hold the primes
    allocate(array(count))
    count = 0

    ! Store the prime numbers in the result array
    do i = 2, limit
      if (sieve(i)) then
        count = count + 1
        array(count) = i
      end if
    end do

  end function primes

end module sieve
